osclib/remote_project.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import re
+
+from lxml import etree as ET
+
+import osc.core
+import osc.conf
+
+import osclib.remote_package
+from osclib.remote_package import RemotePackage
+
+from urllib.error import HTTPError, URLError
+
+class ProjectNotFound(Exception):
+    """Raised when Project is not found on server"""
+    pass
+
+class RemoteProject(object):
+    """This class represents a project on build service side.
+
+    The class offers methods to query and modify the project.
+    The class methods can be used to find or create projects.
+
+    Not to be confused with the class Project in osc.core_, aimed to local checkout of project
+
+    .. _osc.core: https://github.com/openSUSE/osc/blob/master/osc/core.py
+
+    """
+    def __init__(self, name):
+        self.name = name
+        self.metadata = None
+
+    def create_subproject(self, name, title=None, description=None):
+        """Creates subproject with given name. Title and description can be passed,
+        rest of metadata is copied from self."""
+        raw_metadata = ProjectMetadata.load(self.name)
+        fullname = self.name + ":" + name
+        root = ET.fromstring(raw_metadata)
+        root.set('name', fullname)
+        for child in root:
+            if child.tag == "title" and title:
+                child.text = title
+            elif child.tag == "description" and description:
+                child.text = description
+        modified_metadata = ET.tostring(root)
+        ProjectMetadata.save(fullname, modified_metadata)
+
+        result = RemoteProject(fullname)
+        result.metadata = ProjectMetadata.parse(modified_metadata)
+        return result
+
+    def get_packages(self, inherited = False):
+        pkg_list = osc.core.meta_get_packagelist(osc.conf.config['apiurl'], self.name)
+        res = [RemotePackage(pkg_name, project=self.name) for pkg_name in pkg_list]
+        if inherited:
+            return self._merge_inherited(res, self.metadata.linked_projects(recursive=True))
+        else:
+            return res
+
+    def _merge_inherited(self, res, linked_projects):
+        merge_dict = { pkg.name: pkg for pkg in res }
+        cleaning_regex = re.compile(r'\.[0-9]+$')
+        for project in linked_projects:
+            for pkg in project.get_packages(inherited = False):
+                # TODO: is there better way then regex?
```

Comment:
Note that ruby2.5 also matches your regexp but is not maintenance specific, so relying on regexp is basically a no go. `/source/SUSE:SLE-15-SP2:Update/zypper.15590/_meta` will tell you `<releasename>zypper</releasename>`.

Reply:
Good catch with that ruby...in general I a bit worry about performance as when I check SLE15-SP4 there was a lot of updates in update channel, but lets not preoptimalize in advance.